rfd/0109-opsgenie-plugin.md

```diff
@@ -0,0 +1,130 @@
+
+### Executing
+
+The plugin will be started using a command of the form
+
+```
+teleport plugin opsgenie start --config /etc/teleport.yaml
+```
+
+## UX
+
+Once an access request has been created, the Opsgenie plugin will create an alert in the service specified in the request annotation using the Opsgenie Alert API [Create](https://docs.opsgenie.com/docs/alert-api#create-alert) endpoint. 
+
+The appropriate on-call responder can then click the provided link to the access request and approve or deny it.
+
+For auto approval of certain access requests the access request will be auto approved if the requesting user is on-call in one of the services provided in request annotation.
+
+Once an access request has been approved or denied the plugin will add a note to the alert and close the relevant alert tied to that access request.
+
+Example role with services in the annotation that indicate that users with this roll can be on-call for those services.
+
+```
+kind: role
+metadata:
+  name: someRole
+spec:
+  allow:
+    request:
+      roles: [someOtherRole]
+      annotations:
+        opsgenie_services: ["service1", "service2"]
+```
+
+## Implementation details
+In this section we will take a look at how the plugin will interact with the Opsgenie API.
+
+### Authorization
+
+The plugin will use the API key provided in the Opsgenie config file when interacting with the Opsgenie API. This API key will be included in the headers of the requests made.
+
+```
+Header Key: Authorization
+Header Value: genieKey $apiKey
+```
+
+### Creating alerts
+When the Opsgenie plugin creates alerts for incoming access requests the [Create](https://docs.opsgenie.com/docs/alert-api#create-alert) alert request will be of the form
+
+```
+{
+	"message": "Access request from <someuser>",
+	"description":"<someuser> requested permissions for roles <someroles> on Teleport at <someTime>.
+ 	Reason: <someReason>
+ 	To approve or deny the request, proceed to <link to the access request>",
+	"responders":[
+    	....
+	],
+	"tags": ["TeleportAccessRequest"],
+	"priority":"<somePriority>"
+}
+```
+
+On every review of the access request the alert created in Opsgenie will have a note added to it using the ‘[Add note to alert](https://docs.opsgenie.com/docs/alert-api#add-note-to-alert)’ endpoint. Then the alert will be closed using the ‘[Close alert](https://docs.opsgenie.com/docs/alert-api#close-alert)’ endpoint once the alert has either been approved or denied.
+
+```
+<Reviewer> reviewed the request at <someTime>.
+Resolution: <StateEmojiAsUsedByExistingPLugins> <State>.
+Reason: <Reason>
+```
+
+### Auto approval
+
+To check if the requesting user of a request is currently on-call the ‘Who is on call API’s ‘[Get on calls](https://docs.opsgenie.com/docs/who-is-on-call-api#get-on-calls)’ endpoint will be used. 'https://<configured-opsgenie-address>/v2/schedules/<SheduleName>/on-calls?scheduleIdentifierType=name'
```

Comment:
Where do we get the ScheduleName?
Should it be part of the opsgenie config?

Reply:
Had not considered this. We could add it to the config but maybe it would make more sense for this to be part of the request annotations since the service is already there.

Reply:
I think a default value in config and then request annotations if override is necessary

Well, up to you where we put them :+1: